File: parse.pl
use warnings;
use strict;
use HTML::TreeBuilder::XPath;
use Data::Dumper;
use JSON::PP;
#use HTML::Entities;

#binmode STDOUT, ':encoding(UTF-8)';

my @fns = @ARGV;

my @dialogs;

foreach my $fn (@fns) {
    print STDERR $fn, "\n";
    my $xp = HTML::TreeBuilder::XPath->new;

    $xp->parse_file($fn);

    my $dialog;

    my $title;
    if ($xp->exists('//div[@class="head2"]/p')) {
	$title = $xp->findnodes('//div[@class="head2"]/p')->[0]->as_trimmed_text;
    } elsif ($xp->exists('//p[@class="head2"]')) {
	$title = $xp->findnodes('//p[@class="head2"]')->[0]->as_trimmed_text;
    } elsif ($xp->exists('//div[@class="head1"]/p')) {
	$title = $xp->findnodes('//div[@class="head1"]/p')->[0]->as_trimmed_text;
    } elsif ($xp->exists('//p[@class="head1"]')) {
	$title = $xp->findnodes('//p[@class="head1"]')->[0]->as_trimmed_text;
    } elsif ($xp->exists('//div[@class="head1"]')) {
	$title = $xp->findnodes('//div[@class="head1"]')->[0]->as_trimmed_text;
    } elsif ($xp->exists('//div[@class="head2"]')) {
	$title = $xp->findnodes('//div[@class="head2"]')->[0]->as_trimmed_text;
    }
    
    if (not defined $title) {
	warn "No title found in $fn";
    }
    # trim
#    $title =~ s/^\s+//g;
#    $title =~ s/\s+$//g;
    $dialog->{title} = $title;
    
    # remove all <span class="dir">....</span> from doc
    foreach my $stage_directions ($xp->findnodes('//p/span[@class="dir"]'), $xp->findnodes('//span[@class="dir"]')) {
	$stage_directions->detach;
    }

    my @parsed_lines;
    
    # all p's that contain a <span class="dpart">
    # that works for every show EXCEPT princess ida
    my @lines = $xp->findnodes('//p[./span[@class="dpart"]]');
    if (@lines) {
	foreach my $line (@lines) {
	    # get the part name
	    my $part_name = $line->findvalue('span[@class="dpart"]');

	    # clean up part name
	    $part_name =~ s/^\s+//g;
	    $part_name =~ s/\s+$//g;
	    $part_name =~ s/\.//g;
	
	    # remove the part name from the line
	    foreach ($line->findnodes('span[@class="dpart"]')) { $_->detach };

	    my $line_obj = { part => $part_name, line => $line->as_trimmed_text };
	    push @parsed_lines, $line_obj;

	}
    } else {
	my @part_rows = $xp->findnodes('//tr[./td[@class="part"]]');
	my $previous_part;
	my @previous_lines;
	foreach my $row (@part_rows) {
	    my $part_name = $row->findnodes('td[@class="part"]')->[0]->as_trimmed_text;
	    if ($part_name !~ /\w/) {
		# continuation of the previous line, yay! omg
		$part_name = $previous_part;
	    }
		
	    $part_name =~ s/\.//g;
	    my $line = join " ", grep { /\w/ } map { $_->as_trimmed_text } $row->findnodes('td[@class="tlyric"]');
	    
	    if (not defined $previous_part or $part_name ne $previous_part) {
		if (@previous_lines) {
		    my $previous_line_obj = { part => $previous_part, line => join " ", @previous_lines };
		    push @parsed_lines, $previous_line_obj;
		}
		@previous_lines = ();
	    }

	    $previous_part = $part_name;
	    push @previous_lines, $line;
	}
	if (@previous_lines) {
	    my $previous_line_obj = { part => $previous_part, line => join " ", @previous_lines };
	    push @parsed_lines, $previous_line_obj;
	}	      

    }

    $dialog->{lines} = \@parsed_lines;
    push @dialogs, $dialog;
    
#    print Dumper(\@lines);

}

#print Dumper(\@dialogs);
print encode_json(\@dialogs);
    

File: parse.py
from lxml import html
from lxml import etree
import sys
import re

def as_trimmed_text(node):
    res = ""
    for t in node.xpath(".//text()"):
        res += t.strip()
    return res


dialogs = []
fns = sys.argv[1:]

for fn in fns:
    fh = open(fn)
    xp = html.fromstring(fh.read())
    fh.close()

    dialog = dict()

    title = None

    if xp.xpath('//div[@class="head2"]/p'):
        title = xp.xpath('//div[@class="head2"]/p')[0].text_content()
    elif xp.xpath('//p[@class="head2"]'):
        title = xp.xpath('//p[@class="head2"]')[0].text_content()
    elif xp.xpath('//div[@class="head1"]/p'):
        title = xp.xpath('//div[@class="head1"]/p')[0].text_content()
    elif xp.xpath('//p[@class="head1"]'):
        title = xp.xpath('//p[@class="head1"]')[0].text_content()
    elif xp.xpath('//div[@class="head1"]'):
        title = xp.xpath('//div[@class="head1"]')[0].text_content()
    elif xp.xpath('//div[@class="head2"]'):
        title = xp.xpath('//div[@class="head2"]')[0].text_content()

    if title is None:
        sys.stderr.print("No title found in ", fn)

    print(title)

    # remove all <span class="dir">....</span> from doc
    stage_directions = xp.xpath('//p/span[@class="dir"]') + xp.xpath('//span[@class="dir"]') 
    print(stage_directions)
    for sd in stage_directions:
        if sd.getparent() is not None:
            sd.drop_tree()
    

    parsed_lines = []
    
    # all p's that contain a <span class="dpart">
    # that works for every show EXCEPT princess ida
    lines = xp.xpath('//p[./span[@class="dpart"]]')
    if lines:
        for line in lines:
	    # get the part name
            part_name = line.findall('span[@class="dpart"]')[0].text_content()

	    # clean up part name
            part_name = re.sub(r'^\s+', '', part_name)
            part_name = re.sub(r'\s+$', '', part_name)
            part_name = re.sub(r'\.', '', part_name)
	
	    # remove the part name from the line
            for occ in (line.findall('span[@class="dpart"]')):
                occ.drop_tree()

            line_obj = { 'part': part_name, 'line': as_trimmed_text(line) }
            parsed_lines.append(line_obj)
    else:
        part_rows = xp.xpath('//tr[./td[@class="part"]]')
        previous_part = None
        previous_lines = []
        for row in part_rows:
            part_name = row.findall('td[@class="part"]')[0].text_content()
            if (not re.search(r'\w', part_name)):
                # continuation of the previous line, yay! omg
                part_name = previous_part;
		
            part_name = re.sub(r'\./', '', part_name)
            line = " ".join([ as_trimmed_text(n) for n in row.findall('td[@class=tlyric"]') if re.search(r'\w', n) ])
	    #my $line = join " ", grep { /\w/ } map { $_->as_trimmed_text } $row->findnodes('td[@class="tlyric"]');
	    
            if (not previous_part or part_name != previous_part):
                if previous_lines:
                    previous_line_obj = { 'part': previous_part, 'line': " ".join(previous_lines) }
                    parsed_lines.append(previous_line_obj)

                previous_lines = []
	    

            previous_part = part_name
            previous_lines.append(line)
	
        if (previous_lines):
            previous_line_obj = { 'part': previous_part, 'line': " ".join(previous_lines) }
            parsed_lines.append(previous_line_obj)

    dialog['lines'] = parsed_lines
    dialogs.append(dialog)
    
#    print Dumper(\@lines);

print(dialogs)
'''
#print Dumper(\@dialogs);
print encode_json(\@dialogs);
    
'''

